### ISSUE TITLE
Allow setting the working directory for commands

### ISSUE THREAD

**Issue by bjorn** [bjorn/tiled#941]
It would be nice if the working directory could be specified for the custom commands.

Suggested magic variables:
- `%executablePath` - the path of the executable that's supposed to be run
- `%mapPath` - the path of the current map file

Originally requested at issue #940.

**Comment by IMMZ** [comment]
Implementation is on the way.

**Questions**:
1. @bjorn, How to get '%executablepath%' if user only specified binary file without full path (for example, 'geany %mapfile')?
2. Has 'xterm' some parameter to specify working dir? Can't google some...

**Comment by bjorn** [comment]
> How to get '%executablepath%' if user only specified binary file without full path (for example, 'geany %mapfile')?

One way would be to search the directories in `PATH` for that executable, but this is tricky and depends on the operating-system. So I'd suggest that in this case, `%executablePath` is simply an empty string.

> Has 'xterm' some parameter to specify working dir? Can't google some...

Why would you need this? You're supposed to set the working directory using [`QProcess::setWorkingDirectory`](http://doc.qt.io/qt-4.8/qprocess.html#setWorkingDirectory).

**Comment by IMMZ** [comment]
I thought with "QProcess::setWorkingDirectory(...)' xterm itself runs with working directory, not the process set by user, so it's possible it's not true and I didn't touch 'terminal running' code.

**Comment by bjorn** [comment]
> I thought with "QProcess::setWorkingDirectory(...)' xterm itself runs with working directory, not the process set by user, so it's possible it's not true and I didn't touch 'terminal running' code.

That might very well be and depends on the behavior of xterm, but it's outside the scope of Tiled to resolve such an issue.

**Comment by bjorn** [comment]
Hmm, sorry, I just realized there is an "Execute in Terminal" option in the context menu, though it currently only shows up with Qt 4. I'll push a fix to bring it back for Qt 5.

**Comment by bjorn** [comment]
@IMMZ I quickly checked, and the working directory of the process run with `xterm -e` is matching with the working directory in which `xterm` was run. Just check for example the output in the window when doing:

    cd /tmp
    xterm -hold -e pwd

So at least when using `xterm`, there is no problem here.

**Comment by zcabjro** [comment]
Any reason this wouldn't work for map path? (command.cpp)
```C++

QString Command::finalCommand() const
{
    QString finalCommand = command;

    // Perform variable replacement
    if (MapDocument *mapDocument = DocumentManager::instance()->currentDocument()) {
        const QString fileName = mapDocument->fileName();

        finalCommand.replace(QLatin1String("%mapfile"),
                             QString(QLatin1String("\"%1\"")).arg(fileName));

       QFileInfo fileInfo(fileName);
       QString mapPath = fileInfo.absolutePath();
       finalCommand.replace(
           QLatin1String("%mapPath"),
           QString(QLatin1String("\"%1\"")).arg(absPath));

        //...
    }

    return finalCommand;
}
```

**Comment by bjorn** [comment]
@zcabjro That would work for simply adding a `%mapPath` variable, but this task isn't just about adding that variable but about being able to customize the working directory for the command. @IMMZ tried to implement this, but he eventually closed his pull request (#1004).

**Comment by zcabjro** [comment]
Fair enough. I'll open a pull request with this then since it would be useful for running command scripts relative to the map / without absolute paths.

**Comment by winniehell** [comment]
A few years passed and I'm having the same problem as in #258 again. :smile: I feel like `%executablePath` would be a workaround for that situation, so I would be interested in working on it.

@bjorn What is the expected behavior of `%executablePath` inside of Tiled? I don't feel like a path being relative to the Tiled executable is very helpful. Should it be configurable somewhere?

**Comment by ketanhwr** [comment]
Hi @winniehell! This particular issue is under my proposal for GSoC '17, so I'll close this one within the next month. You can make any other suggestions here after going through my proposal once: [Forum Link](http://discourse.mapeditor.org/t/general-improvements-and-bug-fixes-gsoc-17/2354)

**Comment by winniehell** [comment]
@ketanhwr Thank sounds great, thank you! :+1:

**Comment by ketanhwr** [comment]
Do you think it's a good idea to make a new column "Arguments" just like Qt Creator? This way, we can easily solve the `%executablePath` problem too in the working directory. Here's what I have in mind:

![workingdir-1](https://cloud.githubusercontent.com/assets/2878908/26486415/a5aee3d4-4218-11e7-9baf-9c2db984739d.png)

**Comment by bjorn** [comment]
@ketanhwr Yes, that's definitely useful. I think I've mentioned it before, since it also solves the problem that using the "Browse" button to choose the executable currently resets the command line arguments.

**Comment by bjorn** [comment]
Closed by f59b10ce97939e817680695a8356fc837b3d2d26
